# Working With Time Series

Pandas provides powerful tools for working with time-based data, also called time series.

A time series is simply data recorded over time.

Examples:

**hourly temperature**

**daily stock prices**

**monthly sales**

**yearly population**

The 4 Time Concepts in Pandas

Pandas handles time using four main concepts.

### 1. DateTimes

A DateTime represents a specific moment in time.

Example: 2024-01-01 08:00:00

It includes: **date** , **time** , **optional timezone**

It is similar to Python's datetime.datetime.

Example use:

* timestamp of a transaction

* time a sensor recorded a reading


### 2. TimeDeltas

A TimeDelta represents a difference between two times.

Example: **2 hours** , **5 days** , **30 minutes**

Similar to Python's datetime.timedelta.

Example: **end_time - start_time = duration**

Use case:

* time taken for delivery

* duration of a process

### 3. Time Spans

A Time Span represents a range of time with a defined frequency.

Example:

**hourly temperature readings**
**monthly revenue**
**daily website visits**

Instead of a single timestamp, it represents regular intervals.

This is where pandas time series tools become powerful.

### 4. Date Offsets

A DateOffset represents a calendar-aware time shift.

Examples:

**+1 month**
**+1 business day**
**+1 year**

Unlike timedelta, offsets respect calendar rules.

Example:

**Jan 31 + 1 month → Feb 28**

This is useful for:

* financial data

* business calendars

* monthly reporting



## Creating Time Series Data in Pandas

To work with time series data, pandas usually needs:

* data values

* time labels



Creating DataFrames With Time-Series Labels:

we create a pandas DataFrame using the hourly temperature data from a single day. inorder to work with different time series

You can start by creating a list (or tuple, NumPy array, or other data type) with the data values, which will be hourly temperatures given in degrees Celsius:



In [5]:
# we create a variable temperature, which refers to the list of temperature values.
# This list contains 24 temperature readings, representing each hour of a day.

import pandas as pd
import numpy as np

temperature = [ 8.0,  7.1,  6.8,  6.4,  6.0,  5.4,  4.8,  5.0, 9.1, 12.8, 15.3, 19.1, 21.2, 22.1, 22.4, 23.1,
        21.0, 17.9, 15.5, 14.4, 11.9, 11.0, 10.2,  9.1]
 


In [6]:
temp_dt = pd.date_range(start='2019-10-27 00:00:00.0', periods=24, freq='h')
temp_dt

DatetimeIndex(['2019-10-27 00:00:00', '2019-10-27 01:00:00',
               '2019-10-27 02:00:00', '2019-10-27 03:00:00',
               '2019-10-27 04:00:00', '2019-10-27 05:00:00',
               '2019-10-27 06:00:00', '2019-10-27 07:00:00',
               '2019-10-27 08:00:00', '2019-10-27 09:00:00',
               '2019-10-27 10:00:00', '2019-10-27 11:00:00',
               '2019-10-27 12:00:00', '2019-10-27 13:00:00',
               '2019-10-27 14:00:00', '2019-10-27 15:00:00',
               '2019-10-27 16:00:00', '2019-10-27 17:00:00',
               '2019-10-27 18:00:00', '2019-10-27 19:00:00',
               '2019-10-27 20:00:00', '2019-10-27 21:00:00',
               '2019-10-27 22:00:00', '2019-10-27 23:00:00'],
              dtype='datetime64[ns]', freq='h')

### The Four Main Parameters of date_range()

1. start

**The starting date/time**

Example: 2024-01-01 00:00

2. end

**The last date/time**

Example: 2024-01-01 23:00

3. periods

**The number of timestamps you want**

Example: 24 timestamps

4. freq

**The spacing between timestamps**

Common frequencies:

**"D"** → daily

**"W"** → weekly

**"M"** → monthly

**"H"** → hourly

### Important Rules for date_range()

#### Rule 1  Only 3 parameters can be used at once

You cannot use all four parameters together. Pandas calculates the missing one automatically.

Example:

* start + periods + freq or start + end + freq

#### Rule 2  At least two of these must be provided

From these three:

* start
* end
* periods

You must provide at least two. Otherwise pandas cannot generate the sequence.

#### Rule 3  If freq is missing

Pandas will spread the timestamps evenly between start and end.

In [7]:
# in some insatnce it’s convenient to use date-time values as the row labels:
temp = pd.DataFrame(data={'temperature' : temperature}, index = temp_dt)
temp

,temperature
2019-10-27 00:00:00,8.0
2019-10-27 01:00:00,7.1
2019-10-27 02:00:00,6.8
2019-10-27 03:00:00,6.4
2019-10-27 04:00:00,6.0
2019-10-27 05:00:00,5.4
2019-10-27 06:00:00,4.8
2019-10-27 07:00:00,5.0
2019-10-27 08:00:00,9.1
2019-10-27 09:00:00,12.8


## Indexing and Slicing

since we already have a dataframe we can onveniently apply slicing to get just a part of the information:
from this example its shows how to extract the temperatures between 05:00 and 15:00 (5 a.m. and 3 p.m.).

In [8]:
# Before we index and slice part of our temperature datetime we need to create a Pandas Series first

temperature = pd.Series(temperature, index=temp_dt)
temperature

2019-10-27 00:00:00     8.0
2019-10-27 01:00:00     7.1
2019-10-27 02:00:00     6.8
2019-10-27 03:00:00     6.4
2019-10-27 04:00:00     6.0
2019-10-27 05:00:00     5.4
2019-10-27 06:00:00     4.8
2019-10-27 07:00:00     5.0
2019-10-27 08:00:00     9.1
2019-10-27 09:00:00    12.8
2019-10-27 10:00:00    15.3
2019-10-27 11:00:00    19.1
2019-10-27 12:00:00    21.2
2019-10-27 13:00:00    22.1
2019-10-27 14:00:00    22.4
2019-10-27 15:00:00    23.1
2019-10-27 16:00:00    21.0
2019-10-27 17:00:00    17.9
2019-10-27 18:00:00    15.5
2019-10-27 19:00:00    14.4
2019-10-27 20:00:00    11.9
2019-10-27 21:00:00    11.0
2019-10-27 22:00:00    10.2
2019-10-27 23:00:00     9.1
Freq: h, dtype: float64

In [9]:
# Now we slice our time from 5am to 3pm
temperature.loc['2019-10-27 05':'2019-10-27 15']

2019-10-27 05:00:00     5.4
2019-10-27 06:00:00     4.8
2019-10-27 07:00:00     5.0
2019-10-27 08:00:00     9.1
2019-10-27 09:00:00    12.8
2019-10-27 10:00:00    15.3
2019-10-27 11:00:00    19.1
2019-10-27 12:00:00    21.2
2019-10-27 13:00:00    22.1
2019-10-27 14:00:00    22.4
2019-10-27 15:00:00    23.1
Freq: h, dtype: float64

### Resampling and Rolling

**Resampling:** Resampling means changing the frequency of time-series data.

In other words, you group time data into a different time interval and apply an aggregation such as mean, sum, or count.

You are essentially reorganizing data based on time periods.

#### Why do we need to used Resampling

Real-world datasets often come in one frequency but analysis requires another
**Example :**

* converting our Hourly temperature to get our Daily Average temperature using aggregation such as mean
* Converting our daily sales to monthly revenue using aggregation


**Rolling :** Rolling means calculating statistics over a moving window of data.

Instead of grouping data into fixed periods, it slides across the data step-by-step. Each step calculates a value based on a window of previous observations.

#### Why Rolling is Used 

Rolling helps to:

* smooth noisy data

* detect trends

* compute moving averages

* analyze momentum

It is extremely common in:

* finance

* weather analysis

* sensor data

* forecasting


If you want to split a day into four six-hour intervals and get the mean temperature for each interval
pandas provides the method .resample(), which you can combine with other methods such as .mean():

In [10]:
temp.resample(rule = '6h').mean()

,temperature
2019-10-27 00:00:00,6.616667
2019-10-27 06:00:00,11.016667
2019-10-27 12:00:00,21.283333
2019-10-27 18:00:00,12.016667


Each row corresponds to a single six-hour interval. For example, the value 6.616667 is 
the mean of the first six temperatures from the DataFrame temp, whereas 12.016667 is 
the mean of the last six temperatures.

In [13]:
# To get the minimum temperatures for each interval. You can also use 
temp.resample(rule = '6h').min()

,temperature
2019-10-27 00:00:00,5.4
2019-10-27 06:00:00,4.8
2019-10-27 12:00:00,17.9
2019-10-27 18:00:00,9.1


In [14]:
# To get the maximum temperatures for each interval.
temp.resample(rule = '6h').max()

,temperature
2019-10-27 00:00:00,8.0
2019-10-27 06:00:00,19.1
2019-10-27 12:00:00,23.1
2019-10-27 18:00:00,15.5


In [17]:
'''
Rolling window analysis calculates statistics using a small moving group of adjacent rows.
The window moves one row at a time, and pandas recalculates the statistic for each new window using

'''

temp.rolling(window=3).mean()

,temperature
2019-10-27 00:00:00,NaN
2019-10-27 01:00:00,NaN
2019-10-27 02:00:00,7.300000
2019-10-27 03:00:00,6.766667
2019-10-27 04:00:00,6.400000
2019-10-27 05:00:00,5.933333
2019-10-27 06:00:00,5.400000
2019-10-27 07:00:00,5.066667
2019-10-27 08:00:00,6.300000
2019-10-27 09:00:00,8.966667
